### PR TITLE
feat: use single per-migration txn when possible

### PR DESCRIPTION
**Pull request by crepererum**
If all steps can (and want) to run in a transaction block, then wrap the
migration bookkeeping and the migration script into a single
transaction. This way we avoid the dirty state altogether because its
now an "all or nothing" migration.

Note that we still guarantee that there is only a single migration
running at the same time due to the locking mechanism. Otherwise we
would potentially run into nasty transaction failures during schema
modifications.

This is related to #7897 but only fixes / self-heals the "dirty" state
for transaction that can run in transactions. For concurrent index
migrations (which we need in prod) we need to be a bit smarter and this
will be done in a follow-up. However I feel that not leaving half-done
migrations for the cases where it's technically possible (e.g. adding
columns) is already a huge step forward.

### CODE REVIEW

File: iox_catalog/src/migrate.rs [hunk 1]
```diff
@@ -635,30 +713,39 @@ mod tests {
         #[tokio::test]
         async fn test_step_sql_statement_no_transaction() {
```
Comment (crepererum) [inline iox_catalog/src/migrate.rs:714]: This test has technically NOT changed. I just ported it to a slightly more high-level interface (similar to the other tests, using `IOxMigrator` instead of `IOxMigrationStep` directly) so that I can implement the transaction changes more easily.

File: iox_catalog/src/migrate.rs [hunk 2]
```diff
@@ -837,6 +926,74 @@ mod tests {
...
+                in_transaction: true,
+            };
+
+            let test_query = "SELECT COALESCE(SUM(x), 0)::INT AS r FROM t;";
+
+            let migrator = IOxMigrator::new([IOxMigration {
+                version: 1,
+                description: "".into(),
+                steps: steps_broken,
+                checksum: vec![1, 2, 3].into(),
```
Comment (NGA-TRAN) [inline iox_catalog/src/migrate.rs:965]: Can you explain a bit what the role of `checksum` here given `step_broken` seems does almost nothing? There are no step 1, 2, 3, right?

Comment (crepererum) [reply iox_catalog/src/migrate.rs:966]: done

**Review by NGA-TRAN — APPROVED** [review]

File: iox_catalog/src/migrate.rs (hunk 2 above)
Comment (NGA-TRAN) [reply iox_catalog/src/migrate.rs:966]: @crepererum I feel something I wanted to say yesterday but not clear but now I got it formed and wanted to ask you here.
Usually a transaction broken testing would be you insert/update something then something wrong before the commit and the transaction is cancelled/rollbacked. I wonder whether you can have some tests for that. For example, you have step1 to insert something, then step2 you do some rollback or the like. Then you do the check and do not see the insert. Maybe you already have those tests somewhere already? Can we apply it here?

Comment (crepererum) [reply iox_catalog/src/migrate.rs:966]: I think we have something already: the broken step just executes an illegal SQL statement called `foo` and hence rolls back the entire transaction. This is also asserted in this very test here because after the first migration run NONE of the insertions is present (sum=0). After we "fix" the invalid SQL statement, all of them run through (sum=6).
